### PR TITLE
GEM Slice Test Geometry + related Validation Package Changes (92X)

### PR DESCRIPTION
**Pull request by pietverwilligen**
92X version of PR #18649

This PR fixes the GEMGeometryBuilder to be constructed from the CondDB.
This way the GEM Slice Test geometry (5 chambers) in the negative endcap
gets constructed correctly from the database.
Furthermore it adds several fixes to the Validation Package such that it works
with the GEM Slice geometry.

### CODE REVIEW

**Comment by ianna** [comment]
please test

**Comment by pietverwilligen** [comment]
please wait a second. The 91X tests indicated some RunTheMatrix problems

**Comment by pietverwilligen** [comment]
I updated the PR to reflect the changes also made for PR #18649
Please test.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
On 5/10/17 1:22 AM, Piet Verwilligen wrote:
> This PR fixes the GEMGeometryBuilder to be constructed from the CondDB.
> This way the GEM Slice Test geometry (5 chambers) in the negative endcap
> gets constructed correctly from the database.

Is the geometry used in geant to create simhits wrong as well?
Which part is incorrect?
A drawing or a sketch with a comparison would be nice.

**Comment by ianna** [comment]
@slava77 - no, geometry used in G4 is correct. The part which was missing is an interface to GEM reco geometry hierarchy: ring, chamber, super chamber.

**Comment by slava77** [comment]
On 5/10/17 6:43 AM, Ianna Osborne wrote:
> @slava77 <https://github.com/slava77> - no, geometry used in G4 is
> correct. The part which was missing is an interface to GEM reco geometry
> hierarchy: ring, chamber, super chamber.

Hi Yana,

Thank you for the clarification.
So, the GEN-SIM for the GEM slice test is good and valid.

**Comment by pietverwilligen** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by dmitrijus** [comment]
+1

**Comment by perrotta** [comment]
This branch has conflicts that must be resolved: please rebase

**Comment by Dr15Jones** [comment]
+1
just an IOMC/RandomEngine update to add a module to random number list.

**Comment by perrotta** [comment]
This branch has conflicts: please rebase

**Comment by calabria** [comment]
@pietverwilligen can you please resolve the conflict? we need to close this nevernding story

File: Validation/MuonGEMHits/src/GEMTrackMatch.cc [hunk 1]
```diff
@@ -139,24 +139,28 @@ bool GEMTrackMatch::isSimTrackGood(const SimTrack &t)
...
-    pos_ids.push_back(GEMDetId(1,1,1,1,i,useRoll_).rawId());
-    neg_ids.push_back(GEMDetId(-1,1,1,1,i,useRoll_).rawId());
-    phis.push_back(i*10.);
+
+  for(auto it : gem_geom_->chambers()) {
+    if(it->id().region()>0) {
+      pos_ids.push_back(it->id().rawId());
+      edm::LogInfo("GEMTrackMatch")<<"added id = "<<it->id()<<" = "<<it->id().rawId()<<" to pos ids"<<std::endl;
+    }
+    if(it->id().region()<0) {
```
Comment (perrotta) [inline Validation/MuonGEMHits/src/GEMTrackMatch.cc]: else if (it->id().region()<0) {

File: Validation/MuonGEMHits/src/GEMTrackMatch.cc [hunk 2]
```diff
@@ -139,24 +139,28 @@ bool GEMTrackMatch::isSimTrackGood(const SimTrack &t)
...
+      neg_ids.push_back(it->id().rawId());
+      edm::LogInfo("GEMTrackMatch")<<"added id = "<<it->id()<<" = "<<it->id().rawId()<<" to neg ids"<<std::endl;
+      const BoundPlane& bSurface(it->surface());
+      LocalPoint  lCentre( 0., 0., 0. );
+      GlobalPoint gCentre(bSurface.toGlobal(lCentre));
+      int cphi(static_cast<int>(gCentre.phi().degrees()));
+      if (cphi < 0) cphi += 360;
+      phis.push_back(cphi);
```
Comment (perrotta) [inline Validation/MuonGEMHits/src/GEMTrackMatch.cc]: How can you make the pair with pos_ids at L165 if  "phis" is only filled up for the negative part?

**Review by perrotta — COMMENTED** [review]
A couple of issues to look at for when you'll rebase this PR...

**Comment by kpedro88** [comment]
-1
needs rebase

**Comment by calabria** [comment]
@slava77 can you please tell how can run2_GEM_2017_MCTest be used to produce events with the slice test? I believe if we want to use the comma separated eras we need to include it at least in internalUserMods, right?

**Comment by slava77** [comment]
On 5/29/17 9:39 AM, Cesare Calabria wrote:
> @slava77 <https://github.com/slava77> can you please tell how can
> run2_GEM_2017_MCTest be used to produce events with the slice test? I
> believe if we want to use the comma separated eras we need to include it
> at least in internalUserMods, right?

The expected syntax is
cmsDriver.py ...  --era Run2_2017,run2_GEM_2017_MCTest

If it should be declared in an additional file, please do it.

**Comment by calabria** [comment]
mmmm in order to do that I have added run2_GEM_2017_MCTest to internalUserMods in Era.py, it is recognized and added propertly to the cfg, but actually it is not working because the simulation remains unchanged...

**Comment by kpedro88** [comment]
@calabria can you point us to your cfg/work area?

**Comment by calabria** [comment]
well I don't think you can access, but I have generated my cfg's with the following cmsDriver commands:
cmsDriver.py TenMuE_0_200_pythia8_cfi --conditions auto:phase1_2017_realistic -n 10 --era Run2_2017,run2_GEM_2017_MCTest --eventcontent FEVTDEBUG --relval 10000,100 -s GEN,SIM --datatier GEN-SIM --beamspot Realistic50ns13TeVCollision --geometry DB:Extended --fileout file:step1.root --no_exec
 
cmsDriver.py step2 --conditions auto:phase1_2017_realistic -s DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval2017 --datatier GEN-SIM-DIGI-RAW -n 10 --geometry DB:Extended --era Run2_2017,run2_GEM_2017_MCTest --eventcontent FEVTDEBUGHLT --filein file:step1.root --fileout file:step2.root --no_exec

cmsDriver.py step3 --conditions auto:phase1_2017_realistic -n 10 --era Run2_2017,run2_GEM_2017_MCTest --eventcontent FEVTDEBUGHLT,MINIAODSIM,DQM --runUnscheduled -s RAW2DIGI,L1Reco,RECO,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@miniAODDQM --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO --geometry DB:Extended --filein file:step2.root --fileout file:step3.root --no_exec

**Comment by kpedro88** [comment]
A couple of things:
1) There needs to be a separate file `Modifier_run2_GEM_2017_MCTest_cff.py` for `Eras.py` to be able to import the Era correctly. (And then you have to propagate the change to all files that use the Era, like this:
`for i in */*/*/*.py; do sed -i 's/from Configuration.Eras.Modifier_run2_GEM_2017_cff import run2_GEM_2017_MCTest/from Configuration.Eras.Modifier_run2_GEM_2017_MCTest_cff import run2_GEM_2017_MCTest/g' $i; done`)
2) Make sure you recompile after adding the `Configuration/StandardSequences` package

**Comment by calabria** [comment]
ok thanks I will try

**Comment by calabria** [comment]
Ok it works fine now. Piet is too busy to take care of this PR. Can we close this PR and send another one from my branch?
https://github.com/cms-sw/cmssw/compare/master...calabria:for2017SliceTest

**Comment by kpedro88** [comment]
@calabria please make the PR and I will ask @davidlange6 to close this one

**Comment by calabria** [comment]
done thanks
https://github.com/cms-sw/cmssw/pull/19018

**Comment by calabria** [comment]
#19018

**Comment by davidlange6** [comment]
maybe the next time is the charm for this PR?...

**Pull request closed without merging**